language/diem-framework/modules/DiemId.move

```diff
@@ -0,0 +1,172 @@
+address 0x1 {
+
+/// Module managing Diem ID.
+module DiemId {
+
+    use 0x1::Event::{Self, EventHandle};
+    use 0x1::Vector;
+    use 0x1::CoreAddresses;
+    use 0x1::Roles;
+    use 0x1::Errors;
+    use 0x1::Signer;
+
+    /// This resource holds an entity's domain names needed to send and receive payments using diem IDs.
+    struct DiemIdDomains has key {
+        /// The list of domain names owned by this parent vasp account
+        domains: vector<DiemIdDomain>,
+    }
+
+    /// Struct to store the limit on-chain
+    struct DiemIdDomain has drop, store, copy {
+        domain: vector<u8>, // UTF-8 encoded and 63 characters
+    }
+
+    struct DiemIdDomainManager has key {
+        /// Event handle for `domains` added or removed events. Emitted every time a domain is added
+        /// or removed to `domains`
+        diem_id_domain_events: EventHandle<DiemIdDomainEvent>,
+    }
+
+    struct DiemIdDomainEvent has drop, store {
+        /// Whether a domain was added or removed
+        removed: bool,
+        /// Diem ID Domain string of the account
+        domain: DiemIdDomain,
+        /// On-chain account address
+        address: address,
+    }
+
+    const DOMAIN_LENGTH: u64 = 63;
+
+    // Error codes
+    /// DiemIdDomains resource is not or already published.
+    const EDIEM_ID_DOMAIN: u64 = 0;
+    /// DiemIdDomainManager resource is not or already published.
+    const EDIEM_ID_DOMAIN_MANAGER: u64 = 1;
+    /// DiemID domain was not found
+    const EDOMAIN_NOT_FOUND: u64 = 2;
+    /// DiemID domain already exists
+    const EDOMAIN_ALREADY_EXISTS: u64 = 3;
+    /// DiemIdDomains resource was not published for a VASP account
+    const EDIEM_ID_DOMAINS_NOT_PUBLISHED: u64 = 4;
+    /// Invalid domain for DiemIdDomain
+    const EINVALID_DIEM_ID_DOMAIN: u64 = 5;
+
+    spec module {
+        pragma verify = false;
+    }
+
+    public fun create_diem_id_domain(domain: vector<u8>): DiemIdDomain {
```

Comment:
This function should be private -- we don't want anyone else other than this module to be able to create `DiemIdDomain`s
```suggestion
    fun create_diem_id_domain(domain: vector<u8>): DiemIdDomain {
```